src/synapse-plugins/converter-plugin/converter-plugin-backend.vala

```diff
@@ -0,0 +1,372 @@
+                debug ("Dim2 now %i", dim2);
+                for (int i = 0; i < dim2; i++) {
+                    factor2 *= pfactor;
+                    debug ("Factor2 now %f", factor2);
+                }
+            }
+
+            // The two given units must be traceable to the same root with the same dimensionality.
+            if (parent != null &&
+                ultimate_parent1 == parent.uid &&
+                factor1 > 0 && factor2 > 0 &&
+                dim1 == dim2) {
+                debug ("Final factors %g, %g", factor1, factor2);
+                var d = num * factor1 / factor2;
+
+                return ResultData () {
+                    factor = d,
+                    from_description = descr1,
+                    to_description = descr2
+                };
+            }
+
+            return null;
+        }
+
+        private bool check_match (
+            Unit u,
+            string unit_s,
+            string prefix,
+            int dimension,
+            out bool use_prefix,
+            out bool use_dimension) {
+
+            use_prefix = false;
+            use_dimension = false;
+            string[] ids = {u.uid};
+            string[] abbreviations = u.abbreviations.split ("|");
+            foreach (string s in abbreviations) {
+                ids += s;
+            }
+
+            ids += _(u.description);
+
+            var match = unit_s;
+            // Test match whole unit
+            foreach (string id in ids) {
+                if (match == id) {
+                    debug ("whole unit matches");
+                    return true;
+                }
+            }
+
+            if (prefix != "") {
+                //Test match without prefix
+                match = unit_s[prefix.length : unit_s.length];
+                foreach (string id in ids) {
+                    if (match == id) {
+                        debug ("unit less prefix matches");
+                        use_prefix = true;
+                        return true;
+                    }
+                }
+            }
+
+            if (dimension > 1) {
+                //Test match without dimension
+                match = unit_s[0 : -1];
+                foreach (string id in ids) {
+                    if (match == id) {
+                        debug ("unit less dimension matches");
+                        use_dimension = true;
+                        return true;
+                    }
+                }
+            }
+
+            if (prefix != "" && dimension > 1) {
+                //Test match without either prefix or dimension
+                match = unit_s[prefix.length : -1];
+                foreach (string id in ids) {
+                    if (match == id) {
+                        debug ("unit less both matches");
+                        use_prefix = use_dimension = true;
+                        return true;
+                    }
+                }
+            }
```

Comment:
At this point we still do not know what `prefix` and `dimension` are so we have to test.

Reply:
I might misunderstand the code but didn't `get_prefix_and_dimension` already figure out what the `prefix` and `dimension` are?

What if you replace `prefix` and `dimension` with `""` in `unit_s` and only then try to match it to the `ids`? 

Again I probably misunderstand but why are the `use_prefix` and `use_dimension` needed? Can't we just use the prefix/dimension when they are not empty?

Reply:
OK, I'll rethink this tomorrow.

Reply:
OK, so the logic is a little subtle here.  I think we need `use_prefix` because some units e.g. "minute" start with an SI prefix ("m") but it is not actually a prefix and is not used.  We need to distinguish this from cases like "mm" where the prefix "m" *is* used as part of the factor.  I think you may right however that "use_dimension" can be discarded since no units have a uid, abbreviation or description ending in a number (at the moment) so if we find one then it must be used.

Reply:
Just checked through the `units` database and there are one or two obscure units like "mu0" ending in a number but I think the scope of this tool is likely to be restricted to commonly used units and I cannot think of any that end in a number.

Reply:
I have now simplified the check considerably - thanks.

Reply:
Ah I think I understand the problem. The simplification looks great